### DISCUSSION TITLE
Soft-starting the pinging

### DISCUSSION THREAD

**Discussion by dlakelan** [sqm-autorate/sqm-autorate#75] [Algorithm Design]
In addition to soft-starting by dropping the rate to minimum to get an accurate baseline, I think we also need to not ping what is already hundreds and may become thousands of reflectors all at once twice every second for a few seconds. Here's where I'd like to discuss better methods, also this is the first discussion so I'm getting used to the platform.

**Comment by dlakelan** [comment]
In the conductor, at the beginning, we do:

```
    -- Load up the reflectors shared tables
    reflector_data:set("reflector_tables", {
        peers = tmp_reflectors,
        pool = tmp_reflectors
    })
```
which puts all the reflectors in the peers list... This results in us pinging several hundred reflectors twice a second for several seconds until the peer selector has culled it down to 5.

I propose a simple alternative of inserting just a random selection of 40 reflectors... and then re-peer-select more frequently for the first 15 minutes, maybe every minute? Hopefully by doing that we find a sufficient number of "close" reflectors in the first few minutes.

The tradeoff is that the quality of the reflectors in terms of being close to us will be lower for a while since we'll be taking the minimum 10 from whatever random subset we get to select our 5. We could also **seed** the initial set with a few "known good" candidates, like 1.1.1.1 or 9.9.9.9 or whoever so as to ensure we're considering a few strong candidates right from the start.

Another option would be to just use a seeded set of 5 right from the start... and run reselection more often for 15 mins.

**Reply by Fail-Safe** [reply]
I think this sounds like a great place to start addressing this issue. I was actually starting to type up a similar approach where we try something like 20 at a time over the first few minutes and keep only 5-10 of the fastest ones while picking something like 10 more for a next pass.

Starting with your approach sounds like a fine place to start and we can tweak from there 👍

**Reply by Fail-Safe** [reply]
Oh, specifically around seeding several strong candidates, this is essential. Great call there

**Reply by dlakelan** [reply]
Ok, I think what I'll do is seed with 5 "good candidates" since that's easy, then run reselection every 30 seconds for 5 minutes and every minute after that until 15 mins. By then we should have evaluated quite a few reflectors and every 15 mins thereon is probably fine (until we think otherwise).

I'm going to merge my develop/softstart branch with develop/main and then move forward with this soft start approach on that branch, and hopefully have something we can merge back to develop/main along with the develop/smartewma branch and do a point release 0.4.1

Another question though, why is richb seeing huge ping spikes repeatedly every ~ 15 mins on download? Is this a bug we have?

**Reply by dlakelan** [reply]
Also what do you suggest for the seed reflectors?

**Reply by Fail-Safe** [reply]
> I'm going to merge my develop/softstart branch with develop/main and then move forward with this soft start approach on that branch, and hopefully have something we can merge back to develop/main along with the develop/smartewma branch and do a point release 0.4.1

Sounds good. I'm running your updated `develop/smartewma` branch now and no runtime issues that I have come across.

> Another question though, why is richb seeing huge ping spikes repeatedly every ~ 15 mins on download? Is this a bug we have?

Well at the 15 minute mark the re-selection would occur, so is it possible even testing 20 reflectors is stressing his connection?

**Reply by dlakelan** [reply]
How about?
```
    reflector_data:set("reflector_tables", {
        peers = {"9.9.9.9", "8.238.120.14","74.82.42.42",
            "194.242.2.2","208.67.222.222"},
        pool = tmp_reflectors
    })
```
This is (not in order) Quad9, mullvad, OpenDNS, Verizon, and Hurricane Electric... seems like a decent starting seed.

**Reply by Fail-Safe** [reply]
I would consider also 94.140.14.14 (Adguard)

**Reply by dlakelan** [reply]
> 20 reflectors is stressing his connection?

If it really is just 20 then this seems unlikely, I mean, we're talking about 40 packets per second, a few tens of kbps, so unless he's on a dialup modem I don't think so. Though that makes me think maybe he's on a slow connection and needs to adjust his minimum data rate percentage a big higher, because maybe 20% of his connection is in fact dialup speed?

**Reply by Fail-Safe** [reply]
That's a great list, though. I think all should be pretty darn near 100% availability.

**Reply by Fail-Safe** [reply]
> because maybe 20% of his connection is in fact dialup speed?

That's a really good point. I do think he's on a 30/30mbps fiber connection, though. But I may be off-the-mark on my recollection of that

**Reply by dlakelan** [reply]
![image](https://user-images.githubusercontent.com/4332910/148664251-e53c3067-0d00-4a23-b8c3-dc2756a3dc7d.png)
Yeah, based on his graph the minimum speed is more than 5Mbps so no it's not that he's got a crazy slow connection.

**Reply by Fail-Safe** [reply]
Maybe there is a bug going on with the reflector selector. Let me watch it more closely on my side for a little while.

**Reply by dlakelan** [reply]
Ok I actually just pushed to develop/softstart and I think it should work... it'll do reselection every 30 seconds for 40 times and then switch to 15 mins.

However, as many reflectors will be *new* each cycle, this is going to make baselining them harder. The seeds should get a good baseline from the rate soft-start, but after that, the baseline will start out as whatever the first reflection gives ... and if there's bufferbloat we'll have a hard time reliably detecting it for the first 20 minutes

**Reply by dlakelan** [reply]
![image](https://user-images.githubusercontent.com/4332910/148664562-adff7c83-0fd1-4d0e-a63b-d04d26030a27.png)
This does indicate that probably the re-selection phase is maxing out his connection, notice the relative load spikes going to 1 or above.

**Reply by Fail-Safe** [reply]
Alright, I'm going to restart autorate using this develop/softstart branch. FYI, I just rebased it off lua-threads branch because softstart had the old ICMP reflectors file with only around 190 IPv4 reflectors.

I'll let it run for a bit while I do some family stuff and then I'll be back in probably 30-45 minutes with some logs

**Reply by Fail-Safe** [reply]
Oh wow, that's a "softer" start for sure! Much kinder to the CPU. Alright, I'll let it bake for a bit and then I'll be back with some data...

**Reply by dlakelan** [reply]
Yeah, compared to 600pps or so, 12 to 50 or so is a lot softer.

However, can you run a test in which you start a speed test first, THEN fire up the script... try to keep speed tests running the whole time (maybe run one for a while, then fire up a different site, alternate? dunno) I want to see how well it works in the first say 2 minutes under constant load starting right from the beginning.

**Reply by Fail-Safe** [reply]
> Oh wow, that's a "softer" start for sure! Much kinder to the CPU. Alright, I'll let it bake for a bit and then I'll be back with some data...

Here's the results from this first run...

![delaydownecdf](https://user-images.githubusercontent.com/10307870/148666050-6724e339-eb5d-44de-81f6-819902bae079.png)
![delayupecdf](https://user-images.githubusercontent.com/10307870/148666051-2f14c6f6-31f4-4f51-8674-3bb3d7f7c905.png)
![timeseries](https://user-images.githubusercontent.com/10307870/148666053-97628b19-a5b8-44b6-92af-c4c8c8e9fbb6.png)
![downhist](https://user-images.githubusercontent.com/10307870/148666052-1050bd0d-630a-43be-89fa-4a712c9f08a6.gif)
![uphist](https://user-images.githubusercontent.com/10307870/148666054-6a0d6070-98e9-445a-a548-edc1a80cb3fc.gif)

**Reply by Fail-Safe** [reply]
> However, can you run a test in which you start a speed test first, THEN fire up the script... try to keep speed tests running the whole time (maybe run one for a while, then fire up a different site, alternate? dunno) I want to see how well it works in the first say 2 minutes under constant load starting right from the beginning.

Here we go...
![delaydownecdf](https://user-images.githubusercontent.com/10307870/148666214-100b359c-6965-498b-ab91-54dc605da4b4.png)
![delayupecdf](https://user-images.githubusercontent.com/10307870/148666215-20727715-faf4-4d4e-b37b-0837d8c52987.png)
![timeseries](https://user-images.githubusercontent.com/10307870/148666217-f9320e24-e8eb-4276-9077-de491ea8f13b.png)
![downhist](https://user-images.githubusercontent.com/10307870/148666216-830b37e0-f602-473b-be09-42f2ac62c4a8.gif)
![uphist](https://user-images.githubusercontent.com/10307870/148666218-92560217-b9b0-4bd4-95c2-b58da4fab46c.gif)

**Reply by Fail-Safe** [reply]
Going to run another one now...

**Reply by Fail-Safe** [reply]
This had multiple simultaneous downloads and uploads...

![delaydownecdf](https://user-images.githubusercontent.com/10307870/148666506-f7e9e3fc-dbfb-429e-9bcd-aa2971fe9b6f.png)
![delayupecdf](https://user-images.githubusercontent.com/10307870/148666507-e45b6ceb-bfe9-4c9e-bc64-95ca987235ce.png)
![timeseries](https://user-images.githubusercontent.com/10307870/148666510-cec8a8b5-b462-4e69-9fc2-0876a975c528.png)
![downhist](https://user-images.githubusercontent.com/10307870/148666508-f9988994-c50b-4abb-a47b-acf41b0c7ba4.gif)
![uphist](https://user-images.githubusercontent.com/10307870/148666511-fbe21ff2-e05e-4c23-9474-bb7d2fb488d7.gif)

**Reply by dlakelan** [reply]
Those look really good! 

This graph shows my concern about ping spikes. Even when the load goes to zero near the end, especially the up-delay is quite noisy with filtered values in the range 5-10ms varying randomly with little spikes here and there. If we took those raw, I'd be concerned that they might trigger the down-step, and with the load near zero, it would decide to down-step to the minimum.

![image](https://user-images.githubusercontent.com/4332910/148667226-204fb2d2-ab19-4e07-8384-1efb1fbde701.png)

**Reply by Fail-Safe** [reply]
Visually I see what you're talking about, but not sure I'm grasping the big-picture. But if you are, that's great 👍

**Reply by dlakelan** [reply]
Yeah, I don't know the answer, but I think what's going on is that your line has some jitter in its upload latency probably due to shared resources between you and other neighbors, and it means you'll get occasional spikes of delay to 10-15ms. If we don't filter out those short-term delays that last a fraction of a second, then we'll react immediately to those spikes, and cause a dramatic drop in rates. When we do filter that, it makes it take longer before we notice delay, but we don't notice spurious noise. there's a tradeoff there.

I've merged the smartewma stuff with the soft-start onto the develop/softstart branch. Can you pull develop/softstart and run a test similar to one of these, not super long, a few minutes, but with both the soft-start and the smartewma features together? I want to make sure there's no interaction between the two that would be concealed because we had them on separate branches.

**Reply by Fail-Safe** [reply]
![delaydownecdf](https://user-images.githubusercontent.com/10307870/148668243-57b87aa9-7b8b-4840-819e-0e97638636ac.png)
![delayupecdf](https://user-images.githubusercontent.com/10307870/148668244-560ec727-53ca-4f95-b85d-a85427eecd9c.png)
![timeseries](https://user-images.githubusercontent.com/10307870/148668246-8443dd30-2331-4188-9d07-4b34cb6f3d4f.png)
![downhist](https://user-images.githubusercontent.com/10307870/148668245-3379694e-3319-45ab-bbe0-abe72b6ef0c2.gif)
![uphist](https://user-images.githubusercontent.com/10307870/148668247-6a7eda2a-c4f4-4863-8ec3-19dfb8aa944b.gif)

**Reply by Fail-Safe** [reply]
Seemed fine runtime wise. What about the graphs?

**Comment by Fail-Safe** [comment]
@dlakelan  Should we move toward a v0.4.1 merge?

**Reply by dlakelan** [reply]
Yeah those graphs look good, and I like the behavior of the delay graphs. We see at each trigger the delay drops down rapidly, so the controller is doing its thing. That was a pretty big excursion at ~180 seconds though! but I guess that's just your line doing its thing. 

I do like the idea of 0.4.1 merge, is there anything outstanding bugfixy that needs to go in besides this stuff?

**Reply by Fail-Safe** [reply]
No, nothing else I'm aware of. If needed we could always drop another bug fix release (0.4.2) tomorrow.

**Reply by Fail-Safe** [reply]
@dlakelan #79 is ready to merge if you are.

**Comment by dlakelan** [comment]
Rather than merging from develop/softstart -> testing/lua-threads let's merge develop/softstart -> develop/main and then tag version 0.4.2 on develop/main and merge from there to testing/lua-threads, that way main reflects a continuous history?

**Reply by Fail-Safe** [reply]
You are correct, thanks for catching that. I guess 1:00am is catching up to me.

Got that sorted and released: https://github.com/Fail-Safe/sqm-autorate/discussions/81
